daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -642,94 +643,171 @@ private[lf] object Speedy {
+          case TApp(tyfun, arg) => TApp(go(tyfun), go(arg))
+          case _: TForall | _: TStruct | _: TSynApp | _: TVar =>
+            crash(
+              s"Unexpected type $typ application when replacing parameters in value import -- all types should be serializable, and $typ are not."
+            )
+          case _ =>
+            typ
+        }
+
+      go(typ0)
+    }
+
     // This translates a well-typed LF value (typically coming from the ledger)
     // to speedy value and set the control of with the result.
     // All the contract IDs contained in the value are considered global.
     // Raises an exception if missing a package.
-    private[speedy] def importValue(value: V[V.ContractId]): Unit = {
-      def go(value0: V[V.ContractId]): SValue =
-        value0 match {
-          case V.ValueList(vs) => SList(vs.map[SValue](go))
-          case V.ValueContractId(coid) =>
-            addGlobalCid(coid)
-            SContractId(coid)
-          case V.ValueInt64(x) => SInt64(x)
-          case V.ValueNumeric(x) => SNumeric(x)
-          case V.ValueText(t) => SText(t)
-          case V.ValueTimestamp(t) => STimestamp(t)
-          case V.ValueParty(p) => SParty(p)
-          case V.ValueBool(b) => SBool(b)
-          case V.ValueDate(x) => SDate(x)
-          case V.ValueUnit => SUnit
-          case V.ValueRecord(Some(id), fs) =>
-            val values = new util.ArrayList[SValue](fs.length)
-            val names = fs.map {
-              case (Some(f), v) =>
-                values.add(go(v))
-                f
-              case (None, _) => crash("SValue.fromValue: record missing field name")
-            }
-            SRecord(id, names, values)
-          case V.ValueRecord(None, _) =>
-            crash("SValue.fromValue: record missing identifier")
-          case V.ValueVariant(None, _variant @ _, _value @ _) =>
-            crash("SValue.fromValue: variant without identifier")
-          case V.ValueEnum(None, constructor @ _) =>
-            crash("SValue.fromValue: enum without identifier")
-          case V.ValueOptional(mbV) =>
-            SOptional(mbV.map(go))
-          case V.ValueTextMap(entries) =>
-            SGenMap(
-              isTextMap = true,
-              entries = entries.iterator.map { case (k, v) => SText(k) -> go(v) },
-            )
-          case V.ValueGenMap(entries) =>
-            SGenMap(
-              isTextMap = false,
-              entries = entries.iterator.map { case (k, v) => go(k) -> go(v) },
-            )
-          case V.ValueVariant(Some(id), variant, arg) =>
-            compiledPackages.getSignature(id.packageId) match {
-              case Some(pkg) =>
-                pkg.lookupDefinition(id.qualifiedName).fold(crash, identity) match {
-                  case DDataType(_, _, data: DataVariant) =>
-                    SVariant(id, variant, data.constructorRank(variant), go(arg))
+    private[speedy] def importValue(typ0: Type, value0: V[V.ContractId]): Unit = {
```

Comment:
this assumes everything is well-typed right? you seem to do basically no validation. If so, I suggest adding a comment documenting this.